code/modules/storage/storage.dm

```diff
@@ -0,0 +1,413 @@
+		user.swap_hand()
+
+/// storage item is dropped
+/datum/storage/proc/storage_item_on_drop(atom/source, mob/user)
+	src.hud?.update(user)
+
+/// when reaching inside the storage item, check for traps
+/datum/storage/proc/mousetrap_check(mob/user)
+	if (!ishuman(user) || user.stat)
+		return FALSE
+	for (var/obj/item/mousetrap/MT in src.get_contents())
+		if (MT.armed)
+			user.visible_message("<span class='alert'><B>[user] reaches into \the [src.linked_item.name] and sets off a mousetrap!</B></span>",\
+				"<span class='alert'><B>You reach into \the [src.linked_item.name], but there was a live mousetrap in there!</B></span>")
+			MT.triggered(user, user.hand ? "l_hand" : "r_hand")
+			return TRUE
+	for (var/obj/item/mine/M in src.get_contents())
+		if (M.armed && M.used_up != TRUE)
+			user.visible_message("<span class='alert'><B>[user] reaches into \the [src.linked_item.name] and sets off a [M.name]!</B></span>",\
+				"<span class='alert'><B>You reach into \the [src.linked_item.name], but there was a live [M.name] in there!</B></span>")
+			M.triggered(user)
+			return TRUE
+
+/// show storage contents
+/datum/storage/proc/show_hud(mob/user)
+	user.s_active = src.hud
+	src.hud.update(user)
+	user.attach_hud(src.hud)
+
+/// hide storage
+/datum/storage/proc/hide_hud(mob/user)
+	if (user.s_active == src.hud)
+		user.s_active = null
+		user.detach_hud(src.hud)
+
+// ----------------- PUBLIC PROCS ----------------------
+
+/// check if the storage can hold an item or not
+/datum/storage/proc/check_can_hold(obj/item/W)
+	if (!W)
+		return STORAGE_CANT_HOLD
+
+	if (W.cant_drop)
+		return STORAGE_WONT_FIT
+
+	// if can_hold is defined, check against that
+	if (length(src.can_hold))
+		// early skip if weight class is allowed
+		if (src.check_wclass && W.w_class <= src.max_wclass)
+			return STORAGE_CAN_HOLD
+		for (var/type in src.can_hold)
+			if (ispath(type) && istype(W, type))
+				return STORAGE_CAN_HOLD
+		for (var/type in src.can_hold_exact)
+			if (ispath(type) && W.type == type)
+				return STORAGE_CAN_HOLD
+		return STORAGE_CANT_HOLD
+
+	else if (W.w_class > src.max_wclass)
+		return STORAGE_WONT_FIT
+
+	if (src.is_full())
+		return STORAGE_IS_FULL
+
+	return STORAGE_CAN_HOLD
+
+/// when adding an item in
+/datum/storage/proc/add_contents(obj/item/I, mob/user = null, visible = TRUE)
+	if (user?.equipped() == I)
+		user.u_equip(I)
+	src.stored_items += I
+	I.set_loc(src.linked_item)
+	src.hud.add_item(I, user)
+	I.stored = src
+
+	src.add_contents_extra(I, user, visible)
+
+/// available if add_contents needs to be overridden
+/datum/storage/proc/add_contents_extra(obj/item/I, mob/user, visible)
+	// make sure storage item tooltip will be updated
+	if (istype(src.linked_item, /obj/item))
+		var/obj/item/W = src.linked_item
+		W.tooltip_rebuild = TRUE
+	// for storages that change icon with contents
+	src.linked_item.UpdateIcon()
+	// a mob put the item in
+	if (!istype(user))
+		return
```

Comment:
wouldn't this be "something other than a mob put the item in"? also I'm fine just `CRASH`/`stack_trace`ing if this happens since if it does happen we probably want to know

Reply:
Yep, which is okay, since datum/storage/add_contents() is still used to add something to a storage by non-mobs, ex. spawning something in a backpack through job selection

Reply:
Moved comment though https://github.com/goonstation/goonstation/pull/13023/commits/31a1b93405aec5daaaa56399df9416afce8f59ea